File: LC/[1857] Largest Color Value in a Directed Graph.cpp
class Solution {
    int inD[100000];         //n<=10^5
    vector<int>next[100000];
public:
    int largestPathValue(string colors, vector<vector<int>>& edges) 
    {
        int ans = 1; //N>=1
        for (auto edge: edges) //build graph and indegree table
        {            
            int a = edge[0], b = edge[1];
            next[a].push_back(b);
            inD[b]++;
        }
        
        unordered_set<char>Set(colors.begin(), colors.end()); //collect all used colors
        for (char ch='a'; ch<='z'; ch++) //for each color count freq
        {
            if (Set.find(ch)==Set.end()) continue;
            int t = helper(ch-'a', colors, edges);
            if(t==-1) return -1; //check cycle
            ans = max(ans, t);
        }
        return ans;        
    }
    
    int helper(int k, string colors, vector<vector<int>>& edges) 
    {
        int n = colors.size();
        vector<int>count(n, 0); //upto node ii, the count of color k
        vector<int>in(n, 0);    //local indree table
        for (int i=0; i<n; i++) //copy the indree table to local
            in[i] = inD[i];
        
        int nodes = 0;
        queue<int>q; //BFS
        for (int i=0; i<n; i++)
        {
            if (in[i]==0)
            {
                nodes++;
                if (colors[i]-'a'==k) count[i]++;
                q.push({i});
            }
        }
        
        int ret = 0;
        while (!q.empty())
        {
            auto cur = q.front();
            q.pop();            
            
            for (auto p: next[cur])
            {
                count[p] = max(count[p], count[cur]+ (colors[p]-'a'==k));  //count nunber of color k            
                ret = max(ret, count[p]);                
                in[p]--;
                if (in[p]==0)
                {                    
                    nodes++;
                    q.push(p);                 
                }
            }            
        }
        
        if (nodes!=n) return -1;//check cycle
        return ret;
    }
};

File: LC/[1857] Largest Color Value in a Directed Graph.py
###############
# 20231029
###############

class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        """
        - colors = "abaca"
        - colors[i] is a lowercase English letter representing the color of the ith node in this graph
        - edges = [[0,1],[0,2],[2,3],[3,4]]

        => largest color frequency in any path
        -> only consider paths from leaf-to-leaf (maximize the possibility)
        a. build graph: Time/Space: O(N)
        b. for each leaf-node, BFS till end and count for color
        c. return the highest color frequency
        """
        N = len(colors)
        g = collections.defaultdict(set)
        _idt = {ii: 0 for ii in range(N)}

        for a, b in edges:
            g[a].add(b)
            _idt[b] += 1

        colors_set = set(list(colors))

        def bfs(a_color):
            idt = _idt.copy()  # make a local copy of idt since we are modifying it locally
            dq = collections.deque()
            # we cannot rely on parent color cnt to determin the max of curr node color cnt as we might reach current node from diff level
            node2color_freq = defaultdict(int)
            for ii, vv in idt.items():
                if vv == 0:
                    dq.append(ii)  # initialize the queue
                    if colors[ii] == a_color:
                        # initialize the color frequency for those enqueued
                        node2color_freq[ii] += 1
            color_cnt = 0
            node_cnt = 0

            while (sz := len(dq)) > 0:  # BFS until all nodes are visited or encounters cycle
                for _ in range(sz):
                    curr = dq.popleft()
                    node_cnt += 1  # count the node num for cycle detection
                    # when evaluating each curr node, we update the color cnt so far
                    color_cnt = max(color_cnt, node2color_freq[curr])

                    for c in g[curr]:
                        idt[c] -= 1

                        # color count is max(existing count, or parent_count + child_count if child color matches)
                        node2color_freq[c] = max(
                            node2color_freq[c], node2color_freq[curr]+(1 if colors[c] == a_color else 0))

                        if idt[c] == 0:
                            dq.append(c)

            if node_cnt != N:
                return -1
            return color_cnt

        mxf = -N  # since we could return -1, default to max -N.  this might not be necessary as we short-circute the return
        for a_color in colors_set:  # for each color perform topology sort to count max color and detect cycle
            cnt = bfs(a_color)  # BFS each color in order to detect cycle
            if cnt == -1:
                return -1  # short circute when encounter cycle
            mxf = max(mxf, cnt)

        return mxf


###################################
###################################
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        """
        - lowercase English letter (Try letter by letter)
        - color value: highest freq of color on a valid path 
        - return largest color value or -1 if exists cycle
        - for each in scope letter, dfs and count frequency.  capture the highest freq to return
        - use topology sort strategy to detect cycle
        """
        # prepare topology sort
        N = len(colors)
        g = collections.defaultdict(set)
        idt = {ii: 0 for ii in range(N)}
        for a, b in edges:
            g[a].add(b)
            idt[b] += 1

        color_set = set(list(colors))
        mxf = 0

        def bfs(color):
            freq = collections.defaultdict(int)
            idt2 = idt.copy()
            dq = collections.deque()
            for c, v in idt.items():
                if v == 0:
                    dq.append(c)
                    if colors[c] == color:
                        freq[c] += 1

            cnt = 0

            ret = 1
            while (sz := len(dq)) > 0:
                has_color = False
                for _ in range(sz):
                    curr = dq.popleft()
                    cnt += 1

                    for child in g[curr]:
                        idt2[child] -= 1

                        freq[child] = max(
                            freq[child], freq[curr] + (1 if colors[child] == color else 0))
                        ret = max(ret, freq[child])
                        # print(" child: ", child, " freq: ", freq[child], " color: ", color)
                        if idt2[child] == 0:
                            dq.append(child)

            if cnt != N:
                return -1
            return ret
        for color in color_set:

            ans = bfs(color)
            # print(" color: ", color, " ans: ", ans)
            if ans == -1:
                return -1
            mxf = max(mxf, ans)

        return mxf
